common/cuda_hip/matrix/bccoo_kernels.hpp.inc

```diff
@@ -0,0 +1,568 @@
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+
+namespace {
+
+/**
+ * The device function of BCCOO spmv
+ *
+ * @param nnz  the number of nonzeros in the matrix
+ * @param num_blks  the number of blocks in the matrix
+ * @param block_size  the number of nonzeros in each block
+ * @param num_lines  the maximum round of each warp
+ * @param chunk_data  the array where the data are
+ * @param offsets_data  the array where the offset of each block is
+ * @param types_data  the array where the type of each block is
+ * @param cols_data  the array where the initial column of each block is
+ * @param rows_data  the array where the initial row of each block is
+ * @param b  the input dense vector
+ * @param b_stride  the stride of the input dense vector
+ * @param c  the output dense vector
+ * @param c_stride  the stride of the output dense vector
+ * @param scale  the function on the added value
+ *
+ * @tparam ValueType  type of values stored in the matrix
+ * @tparam IndexType  type of matrix indexes stored in the structure
+ * @tparam Closure  type of the function used to write the result
+ */
+template <int subwarp_size = config::warp_size, typename ValueType,
+          typename IndexType, typename Closure>
+__device__ void spmv_kernel(const size_type nnz, const size_type num_blks,
+                            const size_type block_size,
+                            const size_type num_lines,
+                            const uint8* __restrict__ chunk_data,
+                            const size_type* __restrict__ offsets_data,
+                            const uint8* __restrict__ types_data,
+                            const IndexType* __restrict__ cols_data,
+                            const IndexType* __restrict__ rows_data,
+                            const ValueType* __restrict__ b,
+                            const size_type b_stride, ValueType* __restrict__ c,
+                            const size_type c_stride, Closure scale)
+{
+    const auto column_id = blockIdx.y;
+    const auto start_blk = blockIdx.x;
+    const auto jump_blk = gridDim.x;
+
+    const auto start_in_blk = threadIdx.y * subwarp_size + threadIdx.x;
+    const auto jump_in_blk = blockDim.y * subwarp_size;
```

Comment:
IndexType could be int32 and int64, but auto will always adjust to int32. Right?

Reply:
Almost. `auto` will be `uint32` because `threadIdx`, `blockIdx` and `gridDim` are all `uint32` internally. However, that might not be enough, as `int64` can hold much more values. You also use `IndexType` in the `for`-loop below (which is correct), so you might as well use the same type for initialization, limit and the increment number.

Reply:
Done!!